cpp/inc/bond/ext/grpc/win_thread_pool.h

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#if !defined(_WIN32) && !defined(WIN32)
+#error This header is Windows-specific
+#endif
+
+#include <sdkddkver.h>
+
+#if _WIN32_WINNT >= 0x0600   // Thread pool API is available on Windows Vista+
+
+#include <bond/core/config.h>
+
+#include <boost/optional.hpp>
+
+#include <windows.h>        // TODO: Avoid including windows.h in public ones
+
+#include <memory>
+#include <system_error>
+#include <type_traits>
+
+namespace bond { namespace ext { namespace gRPC
+{
+    namespace detail
+    {
+        BOND_NORETURN void Win32Exception(const char* message)
+        {
+            throw std::system_error{
+                std::error_code{ static_cast<int>(GetLastError()), std::system_category() },
+                message };
+        }
+
+    } // namespace detail
+
+    /// @brief Thread pool implementation based on Windows API.
+    class win_thread_pool
+    {
+    public:
+        /// @brief Creates new instance of thread pool with predefined number of min/max threads
+        ///
+        /// @throws std::system_error on failure.
+        win_thread_pool()
+            : win_thread_pool{ pool_type::new_instance }
+        {}
+
+        /// @brief Creates new instance of thread pool with provided number of min/max threads.
+        ///
+        /// @throws std::system_error on failure.
+        win_thread_pool(uint32_t minThreads, uint32_t maxThreads)
+            : win_thread_pool{ pool_type::new_instance, std::make_pair(minThreads, maxThreads) }
+        {}
+
+        /// @brief Returns default thread pool associated with current process.
+        ///
+        /// @throws std::system_error on failure.
+        static win_thread_pool& get_default()
+        {
+            static win_thread_pool s_default{ pool_type::default_instance };
+            return s_default;
+        }
+
+        /// @brief Schedules a callback for execution.
+        ///
+        /// @param callback: functor object to be scheduled.
+        template <typename Callback>
+        void operator()(Callback&& callback)
+        {
+            _impl->schedule(std::forward<Callback>(callback));
+        }
+
+    private:
+        enum class pool_type
+        {
+            new_instance,
+            default_instance
+        };
+
+        class impl
+        {
+        public:
+            impl(pool_type type, boost::optional<std::pair<uint32_t, uint32_t>> numThreads)
+                : _pool{ nullptr, CloseThreadpool },
```

Comment:
<strike>Is it safe to call `CloseTheadpool` on `NULL` in the error case?</strike>
Answering my own question: yes, as `unique_ptr` won't call the deleter for `nullptr`.